core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -613,30 +721,53 @@ public void run() {
                 TimeUnit.NANOSECONDS);
             return;
           }
+          isFatal = retryPlan.isFatal;
         }
-      }
 
-      if (!hasHedging()) {
-        commitAndRun(substream);
-        if (state.winningSubstream == substream) {
-          masterListener.closed(status, trailers);
+        if (isHedging) {
+          Substream sub = null;
+          synchronized (lock) {
+            activeHedges.remove(substream);
+            if (!isFatal) {
+              if (hasPotentialHedging()) {
```

Comment:
I'm bothered by the complexity involved with `activeHedges` and the synchronization here and the next several lines. I had mentioned previously that I wanted `state` to be passed into `hasPotentialHedging`, to make it obvious it was being used. But there's also interactions between state and createSubstream() and HedgingRunnable. I think it is all "correct", but it relies on non-obvious behaviors and takes too much effort to verify.

Most of the trouble seems to be the "grab the lock to modify activeHedges" in random places and trying to figure out what the guarantees are at those points. For example, things would be more obvious if HedgingRunnable changed activeHedges and latestAttempt without releasing the lock between. But I've not even gotten to observing interactions with `hedgingFrozen` and the rescheduler.